substrate/client/network/sync/src/lib.rs

```diff
@@ -407,6 +407,122 @@ impl<B: BlockT> PeerSyncState<B> {
 	pub fn is_available(&self) -> bool {
 		matches!(self, Self::Available)
 	}
+
+	/// Checks if we can initiate a new download from the peer.
+	pub fn can_download(&self) -> bool {
+		self.is_available() || matches!(self, Self::DownloadingNew(_))
+	}
+}
+
+/// State about the download in progress.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct PeerDownloadState<B: BlockT> {
+	/// peer.common_number when the download was started.
+	common_number: NumberFor<B>,
+
+	/// The range being download [start, end).
+	range: Range<NumberFor<B>>,
+
+	/// The blocks downloaded so far.
+	downloaded: Vec<BlockData<B>>,
+}
+
+impl<B: BlockT> PeerDownloadState<B> {
+	fn new(common_number: NumberFor<B>, range: Range<NumberFor<B>>) -> Self {
+		Self { common_number, range, downloaded: Vec::new() }
+	}
+
+	/// Handles the new blocks received from the peer.
+	/// Returns:
+	/// Ok(true): If done and no more requests need to be issued.
+	/// Ok(false): if more requests need to be issued.
+	/// Err(): On any failures.
+	fn handle_blocks(
+		&mut self,
+		who: &PeerId,
+		mut new_blocks: Vec<BlockData<B>>,
+		request: BlockRequest<B>,
+		collection: &mut BlockCollection<B>,
+	) -> Result<bool, BadPeer> {
+		collection.clear_peer_download(who);
+
+		// Validate the blocks.
+		let start_block = match validate_blocks::<B>(&new_blocks, who, Some(request)) {
+			Ok(Some(start_block)) => start_block,
+			Ok(None) => return Ok(true),
+			Err(err) => return Err(err),
+		};
+
+		// Add the new blocks to the downloaded list.
+		let range_len = (self.range.end - self.range.start).saturated_into::<u32>() as usize;
+		if (self.downloaded.len() + new_blocks.len()) > range_len {
+			return Err(BadPeer(*who, rep::BAD_RESPONSE))
+		}
+		let new_blocks_len = new_blocks.len();
+		new_blocks.append(&mut self.downloaded);
+		self.downloaded.append(&mut new_blocks);
+
+		if self.range.start == (self.common_number + One::one()) &&
+			self.downloaded.len() < range_len
```

Comment:
Comment for this check please. Otherwise it'll be hard to maintain this code because looking at this alone, I don't know what `self.range.start == (self.common_number + One::one())` implies

Reply:
I had comments below, moved it up and expanded it

Reply:
> This check self.range.start == (self.common_number + One::one()) looks like a hackish way to determine if PeerDownloadState is responsible for accumulating blocks, or BlockCollection is responsible for this on the upper level in ChainSync. May be it's possible to incorporate this code into BlockCollection so that not distribute the logic over multiple components?

`BlockCollection` currently does something slightly different: manages the downloaded range (with holes + merging holes) across several peers, which the `PeerDownloadState` manages a range with a single peer. It would be cleaner/less complex to keep this separation. The `PeerDownloadState` is just a staging area before adding the blocks to the collection.

> If for some reasons it's not possible to incorporate this code into BlockCollection, I would introduce a clear separation of the code needed to pass only one chunk to BlockCollection, and the code responsible for holding up the blocks. For example, this can be done by introducing a separate enum variant into PeerSyncState for the latter (DownloadingRange) with comments that it's used specifically for fork downloading in multiple chunks.

This change does the same thing, except it enhances the state maintained by `PeerSyncState::DownloadingNew`, to manage download of new blocks from a single/unified path. I feel that adding separate variant only adds more complexity/code duplication

Reply:
The check `self.range.start == (self.common_number + One::one())` looks confusing and can be replaced with a single flag `download_full_range`, the value of which we know from the creation of `PeerDownloadState`. Otherwise we don't need `self.common_number` for anything. This is mostly equivalent to having a dedicated enum type for the case when we need to download the full range.

Having said that, IMO updating `BlockCollection` / importing logic to trigger import only when the parent hash corresponds to the already imported block (instead of comparison by number) would be less hacky than introducing a special case for downloading a range starting from `common_number`.

Reply:
> The check self.range.start == (self.common_number + One::one()) looks confusing and can be replaced with a single flag download_full_range, the value of which we know from the creation of PeerDownloadState. Otherwise we don't need self.common_number for anything. This is mostly equivalent to having a dedicated enum type for the case when we need to download the full range.

Yeah this looks cleaner, changed it to have a separate enum. The current `DownloadingNew` remains unchanged

> Having said that, IMO updating BlockCollection / importing logic to trigger import only when the parent hash corresponds to the already imported block (instead of comparison by number) would be less hacky than introducing a special case for downloading a range starting from common_number.

Agreed. I looked at it, this looks like a much bigger change to use block hash instead of block number (this potentially touches several other paths). I feel this is better done as a separate change